### PR TITLE
YOLOv5 segmentation now is possible!

### PR DESCRIPTION
**Pull request by hdnh2006**
In this PR you will be able to use YOLOv5 segmentation from Ultralytics to create masks and accelerate the process in labeling data.

All the segmentation models can be downloaded from here: https://github.com/hdnh2006/anylabeling-assets/releases/tag/v0.0.0

They have been exported with `--opset 12` as option as following:
`python export.py --weights yolov5x-seg.pt --include onnx --opset 12`

An important change has been done in the `blob` preprocessing of all yolo models, since the image is already in BGR format, so `swapRB=False` has been set in order to avoid changes in the channels.

Again, if there is any change you need, please let me know.

### CODE REVIEW

File: anylabeling/services/auto_labeling/yolov5_seg.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+from PyQt5 import QtCore
+
+from anylabeling.views.labeling.shape import Shape
+from anylabeling.views.labeling.utils.opencv import qt_img_to_cv_img
+
+from .model import Model
+from .types import AutoLabelingResult
+
+
+class YOLOv5_seg(Model):
```
Comment (vietanhdev) [inline anylabeling/services/auto_labeling/yolov5_seg.py:15]: Please rename `YOLOv5_seg` to `YOLOv5Segmentation` to follow [PEP8](https://peps.python.org/pep-0008/#class-names).

File: anylabeling/services/auto_labeling/yolov5_seg.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            if score >= self.config["score_threshold"]:  # filtering by score_tresh
+            
+                if self.output_mode == "polygon":
+                    # Transform coordinates
+                    left = int(round(box[0] * seg_width / image_width))
+                    top = int(round(box[1] * seg_height / image_height))
+                    width = int(round(box[2] * seg_width / image_width))
+                    height = int(round(box[3] * seg_height / image_height))
+    
+                    # process protos
```
Comment (vietanhdev) [inline anylabeling/services/auto_labeling/yolov5_seg.py:130]: process protos -> Process protos

File: anylabeling/services/auto_labeling/yolov5_seg.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+                    protos = outputs[1][0, :, top : top + height, left : left + width].reshape(seg_channels, -1)
+                    protos = np.expand_dims(masks[i], 0) @ protos  # matmul
+                    protos = 1 / (1 + np.exp(-protos))  # sigmoid
+                    protos = protos.reshape(height, width)  # reshape
+                    mask = cv2.resize(protos, (box[2], box[3]))  # resize mask
+                    mask = mask >= self.config["mask_threshold"] # filtering mask by tresh
+    
+                    # add mask to overlay layer
```
Comment (vietanhdev) [inline anylabeling/services/auto_labeling/yolov5_seg.py:138]: # Add mask to overlay layer

**Review by vietanhdev — CHANGES_REQUESTED** [review]
Thank you for your contribution.
I think you have skipped all pre-commit when committing your code. 
- Please run `black -l 79 .` to format your code.
- Run `pre-commit run --all-files` and refactor your code to follow this repo's standards.

image

**Review by hdnh2006 — COMMENTED** [review]
All changes requested have been done

File: anylabeling/services/auto_labeling/yolov5_seg.py (hunk 1 above)
Comment (hdnh2006) [reply anylabeling/services/auto_labeling/yolov5_seg.py]: Done

File: anylabeling/services/auto_labeling/yolov5_seg.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+                        ]
+                        avg_area = np.mean(areas)
+
+                        filtered_approx_contours = [
+                            contour
+                            for contour, area in zip(approx_contours, areas)
+                            if area > avg_area * 0.2
+                        ]
+                        contours = filtered_approx_contours
+
```
Comment (vietanhdev) [inline anylabeling/services/auto_labeling/yolov5_seg.py:186]: Please add the first point again as the last element to close the contour.
Like the line 271 of this file: <https://github.com/vietanhdev/anylabeling/blob/master/anylabeling/services/auto_labeling/segment_anything.py>
```
points.append(points[0])
```
The contours will not be closed if you don't do it.

image

Comment (vietanhdev) [reply anylabeling/services/auto_labeling/yolov5_seg.py:186]: https://user-images.githubusercontent.com/18329471/234012461-9be02cb9-91a7-4456-8a6d-bb9d6ba8c8e2.png

**Review by vietanhdev — CHANGES_REQUESTED** [review]

**Comment by hdnh2006** [comment]
Thanks @vietanhdev, 

Please, could you share the image you are using? so both, can work on the same source.

**Comment by vietanhdev** [comment]
Please check this image.
[photo-1595950009887-e9842bcbc1ae.png.zip](https://github.com/vietanhdev/anylabeling/files/11366362/photo-1595950009887-e9842bcbc1ae.png.zip)

**Pull request closed without merging**